doc/source/data/training_ingest.rst

```diff
@@ -0,0 +1,525 @@
+
+In this case, since Ray supports heterogeneous clusters, you can add more CPU-only nodes to your cluster. This adds more CPU cores to your cluster to further parallelize preprocessing. It also adds more object store memory, allowing more data to be buffered in between preprocessing and training if training is bottlenecked by memory.
+
+If your preprocessing is more compute heavy than memory heavy, then add enough CPU only nodes to :ref:`cache the preprocessed dataset <dataset_cache_performance>`.
+
+Prefetching batches
+###################
+While iterating over your dataset for training, you can increase ``prefetch_batches`` in :meth:`iter_batches <ray.data.DataIterator.iter_batches>` or :meth:`iter_torch_batches <ray.data.DataIterator.iter_torch_batches>` to further increase performance. While training on the current batch, this launches N background threads to fetch and process the next N batches.
+
+This can help if training is bottlenecked on cross-node data transfer or on last-mile preprocessing such as converting batches to tensors or executing ``collate_fn``. However, the tradeoff is that increasing ``prefetch_batches`` leads to more data that needs to be held in heap memory. By default, ``prefetch_batches`` is set to 1.
+
+For example, the following code prefetches 10 batches at a time for each training worker:
+
+.. testcode::
+    import ray
+    from ray.air import ScalingConfig, session
+    from ray.train.torch import TorchTrainer
+
+    ds = ray.data.read_text(
+        "s3://anonymous@ray-example-data/sms_spam_collection_subset.txt"
+    )
+
+    def train_loop_per_worker():
+        # Get an iterator to the dataset we passed in below.
+        it = session.get_dataset_shard("train")
+        for _ in range(2):
+            # Prefetch 10 batches at a time.
+            for batch in it.iter_batches(batch_size=128, prefetch_batches=10):
+                print("Do some training on batch", batch)
+
+    my_trainer = TorchTrainer(
+        train_loop_per_worker,
+        scaling_config=ScalingConfig(num_workers=2),
+        datasets={"train": ds},
+    )
+    my_trainer.fit()
+
+.. testoutput::
+    :hide:
+
+    ...
+
+Random Shuffling
+~~~~~~~~~~~~~~~~
+Randomly shuffling data for each epoch can be important for model quality depending on what model you are training.
+
+With Ray Data, there are 2 approaches to random shuffling:
+
+1. Shuffling data blocks & local shuffling on each training worker.
+2. Full global shuffle, which is more expensive.
+
+For most cases, option 1 will suffice. 
+
+First, randomize each :ref:`block <dataset_concept>` of your dataset via :meth:`randomize_block_order <ray.data.Dataset.randomize_block_order>`. Then, when iterating over your dataset during training, enable local shuffling by specifying a ``local_shuffle_buffer_size`` to :meth:`iter_batches <ray.data.DataIterator.iter_batches>` or :meth:`iter_torch_batches <ray.data.DataIterator.iter_torch_batches>`.
```

Comment:
Same comment as above, when do I use iter_batches and when do I use iter_torch_batches?